bitlist64.go

```diff
@@ -0,0 +1,320 @@
+package bitfield
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math/bits"
+)
+
+var _ = Bitfield(&Bitlist64{})
+
+const (
+	// wordSize configures how many bits are there in a single element of bitlist array.
+	wordSize = uint64(64)
+	// wordSizeLog2 allows optimized division by wordSize using right shift (numBits >> wordSizeLog2).
+	// Note: log_2(64) = 6.
+	wordSizeLog2 = uint64(6)
+	// bytesInWord defines how many bytes are there in a single word i.e. wordSize/8.
+	bytesInWord = 8
+	// bytesInWordLog2 = log_2(8)
+	bytesInWordLog2 = 3
+	// allBitsSet is a word with all bits set.
+	allBitsSet = uint64(0xffffffffffffffff)
+)
+
+// Bitlist64 is a bitfield implementation backed by an array of uint64.
+type Bitlist64 struct {
+	size uint64
+	data []uint64
+}
+
+// NewBitlist64 creates a new bitlist of size N.
+func NewBitlist64(n uint64) *Bitlist64 {
+	return &Bitlist64{
+		size: n,
+		data: make([]uint64, numWordsRequired(n)),
+	}
+}
+
+// NewBitlist64From creates a new bitlist for a given uint64 array.
+func NewBitlist64From(data []uint64) *Bitlist64 {
+	return &Bitlist64{
+		size: uint64(len(data)) * wordSize,
+		data: data,
+	}
+}
+
+// BitAt returns the bit value at the given index. If the index requested
+// exceeds the number of bits in the bitlist, then this method returns false.
+func (b *Bitlist64) BitAt(idx uint64) bool {
+	// Out of bounds, must be false.
+	if idx >= b.size {
+		return false
+	}
+
+	i := uint64(1 << (idx % wordSize))
+	return b.data[idx>>wordSizeLog2]&i == i
+}
+
+// SetBitAt will set the bit at the given index to the given value.
+// If the index requested exceeds the number of bits in the bitlist, then this method returns false.
+func (b *Bitlist64) SetBitAt(idx uint64, val bool) {
+	// Out of bounds, do nothing.
+	if idx >= b.size {
+		return
+	}
+
+	bit := uint64(1 << (idx % wordSize))
+	if val {
+		b.data[idx>>wordSizeLog2] |= bit
+	} else {
+		b.data[idx>>wordSizeLog2] &^= bit
+	}
+}
+
+// Len returns the number of bits in a bitlist (note that underlying array can be bigger).
+func (b *Bitlist64) Len() uint64 {
+	return b.size
+}
+
+// Bytes returns underlying array of uint64s as an array of bytes.
+// The leading zeros in the bitlist will be trimmed to the smallest byte length representation of
+// the bitlist. This may produce an empty byte slice if all bits were zero.
+func (b *Bitlist64) Bytes() []byte {
+	if len(b.data) == 0 {
+		return []byte{}
+	}
+
+	buf := bytes.NewBuffer(make([]byte, 0, len(b.data)*bytesInWord))
+	if err := binary.Write(buf, binary.LittleEndian, b.data); err != nil {
+		return []byte{}
+	}
+	ret := buf.Bytes()
+
+	// Clear any leading zero bytes.
+	allLeadingZeroes := 0
+	for i := len(b.data) - 1; i >= 0; i-- {
+		leadingZeroes := bits.LeadingZeros64(b.data[i])
```

Comment:
this can be actually much more efficient. given `data[i]` is a `uint64` we can just check 
```
if b.data[i] == 0 
```
this allows us to quickly determine the number of leading zeroes rather than using the more expensive
```
bits.LeadingZeros64
```
in the event the word is `0x0` This should save a few cpu cycles.